### ISSUE TITLE
SFTP - improve use of ssh-agent

### ISSUE THREAD

**Issue by prettySongbird** [rclone/rclone#7011]
#### What is the problem you are having with rclone?

When using rclone with public key authentication and the new OpenSSH format, using rclone directly will fail even if key_use_agent: true is set and pubkey_file is used as well. When the ssh-agent has already cached the corresponding key, rclone works just fine.

So to use rclone with any modern OpenSSH setup, the user needs to manually run ssh-add /path/to/keyfile first.

This could easily be automated through rclone when the keyfile is already set, by running ssh-add automatically when it is available. 

#### What is your rclone version (output from `rclone version`)

rclone v1.62.2
- os/version: arch "rolling" (64 bit)
- os/kernel: 6.3.2 (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.20.4
- go/linking: dynamic
- go/tags: none

#### Which cloud storage system are you using? (e.g. Google Drive)

SFTP OpenSSH_8.4p1 Debian-5+deb11u1, OpenSSL 1.1.1n 

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)

rclone lsd mysshserver:

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)

```
<7>DEBUG : rclone: Version "v1.62.2" starting with parameters ["rclone" "-vv" "lsd" "lemantis-sftp:/"]
<7>DEBUG : rclone: systemd logging support activated
<7>DEBUG : Creating backend with remote mysshserver:
<7>DEBUG : pacer: low level retry 1/10 (error couldn't connect SSH: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain)
...
<7>DEBUG : pacer: low level retry 10/10 (error couldn't connect SSH: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain)
Failed to create file system for "mysshserver:/": NewFs: couldn't connect SSH: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
```
Config:

```
- type: sftp
- host: mysshserver.tld
- user: root
- port: 122
- key_use_agent: true
- shell_type: unix
- md5sum_command: md5sum
- sha1sum_command: sha1sum
- pubkey_file: /home/user/.ssh/mykey.pub


```

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by animosity22** [comment]
Same as the other.

**Comment by ncw** [comment]
When remove my keys from my ssh agent and try to ssh to somewhere, ssh asks for the password and doesn't add it to the ssh-agent.

Is there a mode which ssh does add the identity to the ssh-agent because I think that is what you are suggesting for rclone?

**Comment by prettySongbird** [comment]
> When remove my keys from my ssh agent and try to ssh to somewhere, ssh asks for the password and doesn't add it to the ssh-agent.
> 
> Is there a mode which ssh does add the identity to the ssh-agent because I think that is what you are suggesting for rclone?

Not sure I understand what you mean, but in general you add you ssh keys to the agent by adding "AddKeysToAgent yes" to ~/.ssh/config

**Comment by ncw** [comment]
I've not seen `AddKeysToAgent` before - useful!

So you are suggesting for rclone to implement this functionality from openssh

     AddKeysToAgent
             Specifies whether keys should be automatically added to a running
             ssh-agent(1).  If this option is set to yes and a key is loaded
             from a file, the key and its passphrase are added to the agent
             with the default lifetime, as if by ssh-add(1).  If this option
             is set to ask, ssh(1) will require confirmation using the
             SSH_ASKPASS program before adding a key (see ssh-add(1) for de‐
             tails).  If this option is set to confirm, each use of the key
             must be confirmed, as if the -c option was specified to
             ssh-add(1).  If this option is set to no, no keys are added to
             the agent.  Alternately, this option may be specified as a time
             interval using the format described in the TIME FORMATS section
             of sshd_config(5) to specify the key's lifetime in ssh-agent(1),
             after which it will automatically be removed.  The argument must
             be no (the default), yes, confirm (optionally followed by a time
             interval), ask or a time interval.

This would probably need to be implemented in the upstream library https://github.com/xanzy/ssh-agent - I'm not sure.

**Comment by prettySongbird** [comment]
> So you are suggesting for rclone to implement this functionality from openssh

Not really. What I had in mind was simply to run something like "ssh-add /path/to/keyfile before trying to connect with rclone.
This will add the key to the agent and as for the password.
Afterwards, rclone can use that key for the connection, independent of the key format, since the key no longer has to be handled by rclone directly but instead the authentication is handled by ssh-agent.

In other words:
If I connect to my server using rclone ls myserver:/
I get:
`NewFs: couldn't connect SSH: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain`

However if I first execute `ssh-add /path/to/my/key` and the use rclone, everything works fine.

This could be automated by having the key added to the agent by rclone automatically by running ssh-add
